src/accountsdb/cache.zig

```diff
@@ -0,0 +1,209 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const builtin = @import("builtin");
+const zstd = @import("zstd");
+
+const Account = sig.core.Account;
+const Pubkey = sig.core.pubkey.Pubkey;
+const LruCacheCustom = sig.common.lru.LruCacheCustom;
+const ReferenceCounter = sig.sync.reference_counter.ReferenceCounter;
+
+/// Stores read-only in-memory copies of commonly used *rooted* accounts
+pub const AccountsCache = struct {
+    lru: LRU,
+    allocator: std.mem.Allocator,
+
+    /// Atomically refcounted account
+    pub const CachedAccount = struct {
+        account: Account,
+        ref_count: ReferenceCounter,
+
+        pub fn init(allocator: std.mem.Allocator, account: Account) error{OutOfMemory}!CachedAccount {
+            return .{
+                .account = try account.clone(allocator),
+                .ref_count = .{},
+            };
+        }
+
+        pub fn deinit(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            self.account.deinit(allocator);
+            self.* = undefined;
+        }
+
+        pub fn releaseOrDestroy(self: *CachedAccount, allocator: std.mem.Allocator) void {
+            if (self.ref_count.release()) {
+                self.deinit(allocator);
+                allocator.destroy(self);
+            }
+        }
+
+        // satisfies the type LruCacheCustom expects
+        pub fn releaseOrDestroyDoublePtr(self: **CachedAccount, allocator: std.mem.Allocator) void {
+            releaseOrDestroy(self.*, allocator);
+        }
+
+        pub fn copyRef(self: *CachedAccount) *CachedAccount {
+            if (self.ref_count.acquire()) {
+                return self;
+            } else {
+                @panic("Attempted to acquire reference to destroyed CachedAccount");
+            }
+        }
+
+        fn getMeta(self: *CachedAccount) CachedAccountMeta {
+            return .{
+                .ref_count = self.ref_count,
+            };
+        }
+    };
+
+    /// A copy of a CachedAccout's fields, without the account
+    const CachedAccountMeta = struct {
+        ref_count: ReferenceCounter,
+    };
+
+    pub const LRU = LruCacheCustom(
+        .locking,
```

Comment:
This is redundant with the RwMux. I get that you want internal locking for this structure eventually, but specifying `locking` here isn't good enough for that, which is why you need the RwMux. It's not clear to me how a `locking` Lru could be beneficial here, so I would just make this `non_locking` for now. If you want to start moving towards internal locking, you can support it fully by integrating a RwLock in AccountsCache, and get rid of the RwMux wrapper.